File: Data Structures & Algorithms/trapping-rain-water/submission-1.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int l = 0;
        int r = height.size() - 1;
        int res = 0;
        int maxl = height[l], maxr = height[r];

        while (l < r)
        {
            if (maxl < maxr)
            {
                l++;
                maxl = max(maxl, height[l]);
                res += maxl - height[l];
            } else
            {
                r--;
                maxr = max(maxr, height[r]);
                res += maxr - height[r];
            }
            
        }
        return res;
    }
};

File: Data Structures & Algorithms/eating-bananas/submission-1.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l, r = 1, max(piles)
        res = r

        while l <= r:
            speed = l + ((r - l) // 2)

            time = 0
            for p in piles:
                time += math.ceil(p / speed)

            if time <= h:
                res = speed
                r = speed - 1
            else:
                l = speed + 1

        return res 